### PR TITLE
support --dry-run in kubectl patch command

### LINKED ISSUE #47172: support --dry-run in kubectl patch command
**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see https://kubernetes.io/docs/tasks/debug-application-cluster/troubleshooting/.):

**Note:** Please file issues for subcomponents under the appropriate repo

| Component | Repo                                                               |
| --------- | ------------------------------------------------------------------ |
| kubectl   | [kubernetes/kubectl](https://github.com/kubernetes/kubectl/issues/new) |
| kubeadm   | [kubernetes/kubeadm](https://github.com/kubernetes/kubeadm/issues/new) |

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.):

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):

**Kubernetes version** (use `kubectl version`):
Client Version: version.Info{Major:"1", Minor:"8+", GitVersion:"v1.8.0-alpha.0.184+3837d951919006", GitCommit:"3837d9519190069c44a453c145869e1fced52f23", GitTreeState:"clean", BuildDate:"2017-06-07T06:07:10Z", GoVersion:"go1.8.1", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"8+", GitVersion:"v1.8.0-alpha.0.184+3837d951919006", GitCommit:"3837d9519190069c44a453c145869e1fced52f23", GitTreeState:"clean", BuildDate:"2017-06-07T06:07:10Z", GoVersion:"go1.8.1", Compiler:"gc", Platform:"linux/amd64"}

**Environment**:
- **Cloud provider or hardware configuration**:
- **OS** (e.g. from /etc/os-release):
- **Kernel** (e.g. `uname -a`):
- **Install tools**:
- **Others**:

**What happened**:
support --dry-run in patch command

**What you expected to happen**:

**How to reproduce it** (as minimally and precisely as possible):

**Anything else we need to know**:

ref to #45712

### PR DESCRIPTION
**Pull request by xilabao**
**What this PR does / why we need it**:

**Which issue this PR fixes**: fixes #47172

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by eparis** [comment]
@k8s-bot ok to test
/release-note-none

File: pkg/kubectl/cmd/patch.go [hunk 1]
```diff
@@ -226,7 +220,13 @@ func RunPatch(f cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []strin
 
 		count++
 
-		originalObjJS, err := runtime.Encode(unstructured.UnstructuredJSONScheme, info.VersionedObject)
+		var obj runtime.Object
+		if info.VersionedObject != nil {
+			obj = info.VersionedObject
+		} else {
+			obj = info.Object
+		}
```
Comment (monopole) [inline pkg/kubectl/cmd/patch.go]: originalObj := determineOriginalObj(info)

File: pkg/kubectl/cmd/patch.go [hunk 2]
```diff
@@ -238,14 +238,24 @@ func RunPatch(f cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []strin
 		if err != nil {
 			return err
 		}
+
+		if !reflect.DeepEqual(obj, targetObj) {
```
Comment (monopole) [inline pkg/kubectl/cmd/patch.go]: if equal (i.e. no patch) is the call to info.Refresh still needed?

File: pkg/kubectl/cmd/patch.go (hunk 1 above)
Comment (xilabao) [inline pkg/kubectl/cmd/patch.go]: > Did this just blow up before if VersionedObject was nil?

`--local` and `--dry-run` can not use together. I think yes. PTAL

**Review by monopole — CHANGES_REQUESTED** [review]

File: pkg/kubectl/cmd/patch.go [hunk 3]
```diff
@@ -281,3 +288,13 @@ func getPatchedJSON(patchType types.PatchType, originalJS, patchJS []byte, gvk s
...
 }
+
+func determineOriginalObj(info *resource.Info) runtime.Object {
+	var obj runtime.Object
+	if info.VersionedObject != nil {
+		obj = info.VersionedObject
+	} else {
+		obj = info.Object
+	}
+	return obj
```
Comment (xilabao) [inline pkg/kubectl/cmd/patch.go]: Done. PTAL @monopole

**Review by monopole — APPROVED** [review]
/lgtm

**Comment by xilabao** [comment]
@monopole @eparis approve?

**Comment by fabianofranz** [comment]
Dupe of https://github.com/kubernetes/kubernetes/pull/45920?

**Comment by xilabao** [comment]
@fabianofranz I think  `--dry-run` is not same with `--local`. maybe it's a dupe of #45920 ?  PTAL

**Comment by fabianofranz** [comment]
@xilabao bot will probably ask for the link to an issue.

**Comment by xilabao** [comment]
@k8s-bot pull-kubernetes-node-e2e test this
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Comment by dims** [comment]
Do we need this for 1.8? please rebase.

**Comment by xilabao** [comment]
rebased @eparis @fabianofranz @monopole ptal

**Comment by eparis** [comment]
/retest
just so the bots will leave us alone!!!

**Comment by eparis** [comment]
@fejta I think the bots are angry...

**Comment by timoreimann** [comment]
I'd love to see this PR crossing the finishing line.

@xilabao, anything in your power you could do to drive the PR to completion?

**Comment by xilabao** [comment]
rebased. @fabianofranz @eparis @monopole @dims PTAL

**Comment by timoreimann** [comment]
Still super interested in seeing this one getting pushed through.

@xilabao any chance for another rebase? :-)

**Comment by timoreimann** [comment]
The PR seems to have gone stale. If nobody objects, then I'll carry it over to a PR of my own.

**Comment by dims** [comment]
@timoreimann +1 i'll close this one out

/close

**Pull request closed without merging**